### ISSUE TITLE
multi stage docker file ignores some stages

### ISSUE THREAD

**Issue by eedwards-sk** [vito/oci-build-task#9]
Not sure if this is a buildkit issue or an issue with this task.

Here is a (stripped down pseudo) Dockerfile that, when built using this task, ends up only building the "build stage" and the "run stage" and ignoring the "test stage" entirely.

When building with docker, it correctly builds all stages.

I suspect either this task or buildkit is not correctly handling the `FROM <earlier stage> as foo` step

```
# build stage
# ==========
FROM python:3.8.0 as build
ENV DOCKER_STAGE=build PYTHONUNBUFFERED=1

WORKDIR /build

RUN do-build

# run stage
# ===========
FROM python:3.8.0-slim as run
ENV DOCKER_STAGE=run

CMD ["foo"]

# test stage
# =========
FROM run as test

WORKDIR /workspace

COPY tests/ tests/

RUN python -m unittest discover

# output stage
# ===========
FROM run
```

**Comment by matthope** [comment]
This sounds like the expected behaviour with buildkit.

**Comment by aledeganopix4d** [comment]
> This sounds like the expected behaviour with buildkit.

Do you have a docs reference to this, please?
Also is there a way to tune that behavior?

**Comment by vito** [comment]
It doesn't appear to be clearly documented anywhere, but this is indeed how `buildkit` works - it carefully keeps tracks of dependencies within the `Dockerfile` to avoid performing redundant work and execute rules in parallel where possible.

There's an upstream issue requesting the ability to execute multiple stages at once, but it was resolved via a higher-level project, `buildx`, which just runs multiple targets: https://github.com/moby/buildkit/issues/609

This actually affects our own resource pipelines, which have the same pattern of running tests as a separate stage within the `Dockerfile`. 🤔 I don't think we should change the *default* behavior, but maybe there should be some way to evaluate multiple targets.

**Comment by vito** [comment]
Investigated this a bit the other day, not seeing a perfect solution yet unfortunately.

### "Solution" 1: just run multiple `tasks`

This would require no changes to `oci-build-task`. Basically just run the task multiple times, once for each desired `$TARGET`. It'll do a lot of redundant building unless you pass the `cache` between the tasks:

```yaml
- task: test
  config:
    # ...
    outputs:
    - name: cache
- task: build
  config:
    # ...
    inputs:
    - name: cache
```

I'm pretty sure this works, but I wouldn't really trust the caching to 100% prevent redundant building. This solution would be convenient for me since I wouldn't have to do anything, but this pattern feels pretty clunky and it would be annoying to implement across many pipelines.

### Solution 2: allow specifying multiple targets to `oci-build-task`

The idea here would be to pass along a set of additional targets to build prior to the final build:

```yaml
params:
  ADDITIONAL_TARGETS: test
```

The additional targets would not produce any outputs - they would just be built to e.g. run tests. Maybe later we could allow exporting these targets as outputs too, but that's not important for stages which just run tests, so let's punt on that for now.

`buildx` takes this approach, but it does it in parallel. I tried just running `buildctl` in parallel for each target and it resulted in jumbled output, so I'm not too keen on that approach. This could probably be fixed by foregoing `buildctl` and talking to the daemon directly and processing the output ourselves, but that's a lot more work.

This solution works decently enough if I just build the separate targets serially using the same `buildkitd` daemon. It even caches things without needing to export/import. As you can see in the [full output](https://gist.github.com/vito/acbb0386aad3b6a2f91def839d9a0447), the first target builds and the second target has nothing to do as everything is `CACHED`. This seems good enough to me.

I did run into one gotcha: if you put the context in `.`, like so...:

```yaml
inputs:
- name: context
  path: .
```

...all sorts of things go bad. This places the outputs and caches in the same directory as the context, which causes lines like `COPY . /src` to not be cached and can also cause cryptic errors like this:

```
#23 ERROR: error writing layer blob: rename /tmp/build/e55deab7/cache/ingest/0d64df4f54665a1888e5b482973571bf70c595af54d61d054fa912efb035af4d/startedat.tmp /tmp/build/e55deab7/cache/ingest/0d64df4f54665a1888e5b482973571bf70c595af54d61d054fa912efb035af4d/startedat: no such file or directory: unknown
------
 > exporting cache:
------
error: failed to solve: rpc error: code = Unknown desc = error writing layer blob: rename /tmp/build/e55deab7/cache/ingest/0d64df4f54665a1888e5b482973571bf70c595af54d61d054fa912efb035af4d/startedat.tmp /tmp/build/e55deab7/cache/ingest/0d64df4f54665a1888e5b482973571bf70c595af54d61d054fa912efb035af4d/startedat: no such file or directory: unknown
```

...so don't do that. Put the context in a separate dir and configure `$CONTEXT` instead.

### Solution 3: native `buildctl` support for building multiple targets

It'd be nice if `buildctl` supported this natively so it could all be run as one dependency graph to solve. I believe that was the original spirit of https://github.com/moby/buildkit/issues/609 but it was closed because folks seem happy enough `buildx bake`. But I don't really want to depend on Docker. 🤷🏻‍♂️

Solution 2 seems good enough, so I think I'll just do that.

**Comment by vito** [comment]
Added support for `$ADDITIONAL_TARGETS`.

Also made it so that if you include an output whose name matches the target, the task will produce the same `image.tar`/`digest` (+ `rootfs/...` and `metadata.json` if `$UNPACK_ROOTFS=true`) to it.

**Comment by vito** [comment]
Tested it out in our resource pipelines and shipped as v0.6.0. 👍🏻
